symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/datafeed/impl/DatafeedServiceV2.java

```diff
@@ -1,36 +1,200 @@
+                throw e;
+            }
+        });
+    }
+
+    private V5Datafeed retrieveDatafeed() throws Throwable {
+        log.debug("Start retrieving datafeed from agent");
+        Retry retry = this.getRetryInstance("Retrieve V5Datafeed");
+        List<V5Datafeed> datafeeds = retry.executeCheckedSupplier(() -> {
+            try {
+                return this.datafeedApi.listDatafeed(authSession.getSessionToken(), authSession.getKeyManagerToken());
+            } catch (ApiException e) {
+                if (e.isUnauthorized()) {
+                    log.info("Re-authenticate and try again");
+                    authSession.refresh();
+                } else {
+                    log.error("Error {}: {}", e.getCode(), e.getMessage());
+                }
+                throw e;
+            }
+        });
+        if (!datafeeds.isEmpty()) {
+            return datafeeds.get(0);
+        }
+        return null;
+    }
+
+    private void readDatafeed() throws Throwable {
+        log.debug("Reading datafeed events from datafeed {}", datafeed.getId());
+        RetryConfig config = RetryConfig.from(this.retryConfig)
+                .retryOnException(e -> {
+                    if (e instanceof ApiException && e.getSuppressed().length == 0) {
+                        ApiException apiException = (ApiException) e;
+                        return apiException.isServerError() || apiException.isUnauthorized() || apiException.isClientError();
+                    }
+                    return false;
+                }).build();
+        Retry retry = this.getRetryInstance("Read Datafeed", config);
+        retry.executeCheckedSupplier(() -> {
+            try {
+                V5EventList v5EventList = this.datafeedApi.readDatafeed(
+                        datafeed.getId(),
+                        authSession.getSessionToken(),
+                        authSession.getKeyManagerToken(),
+                        ackId);
+                this.ackId.setAckId(v5EventList.getAckId());
+                List<V4Event> events = v5EventList.getEvents();
+                if (events != null && !events.isEmpty()) {
+                    this.handleV4EventList(events);
+                }
+            } catch (ApiException e) {
+                if (e.isUnauthorized()) {
+                    log.info("Re-authenticate and try again");
+                    authSession.refresh();
+                } else {
+                    log.error("Error {}: {}", e.getCode(), e.getMessage());
+                    if (e.isClientError()) {
+                        try {
+                            log.info("Try to delete the faulty datafeed");
+                            this.deleteDatafeed();
+                            log.info("Recreate a new datafeed and try again");
+                            this.datafeed = this.createDatafeed();
+                        } catch (Throwable throwable) {
+                            e.addSuppressed(throwable);
```

Comment:
woohh first time I see this being used, what does it do?

Reply:
If a read datafeed fails and return error 400, it means that the datafeed is stale or faulty, in this case, the bot will have to delete the datafeed and create a new one. And these two steps are also in retry mechanism, it means we have a nested retry here. And if the nested retry failed, it should be stop the parent retry instead of trigger the parent retry to do it again. That's why I add a Suppressed throwable to check if the throwable was thrown by nested retry or not.